### PR TITLE
Feat/390 RSR-993 reuse old appid and mac addresses

### PR DESCRIPTION
**Pull request by massifben**

### CODE REVIEW

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java [hunk 1]
```diff
@@ -182,6 +199,29 @@ private Optional<SclReportItem> configureControlBlockNetwork(TCommunication tCom
         return Optional.empty();
     }
 
+    private Map<CbKey, AppIdAndMac> getAppIdsAndMacToReuse(SCL scd, @Nullable TCommunication communicationToReuse) {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:202]:
```suggestion
    private Map<CbKey, AppIdAndMac> getAppIdsAndMacToReuse(SCL scd, List<TSubNetwork> subNetworks) {
```

File: pom.xml [hunk 2]
```diff
@@ -135,6 +135,11 @@
                 <artifactId>jackson-databind</artifactId>
                 <version>${jackson-databind.version}</version>
             </dependency>
+            <dependency>
+                <groupId>jakarta.annotation</groupId>
+                <artifactId>jakarta.annotation-api</artifactId>
+                <version>3.0.0</version>
```
[the remark this answers is not in the document]
Comment (massifben) [reply pom.xml:141]: OK

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java [hunk 3]
```diff
@@ -88,19 +91,29 @@ public void removeAllControlBlocksAndDatasetsAndExtRefSrcBindings(final SCL scl)
 
     @Override
     public List<SclReportItem> configureNetworkForAllControlBlocks(SCL scd, CBCom cbCom) {
+        return configureNetworkForAllControlBlocks(scd, cbCom, null);
```
[the remark this answers is not in the document]
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:94]: OK. No need for "@Nullable" then.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java (hunk 1 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:202]: OK

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java [hunk 4]
```diff
@@ -121,7 +134,7 @@ private Stream<SclReportItem> configureNetworkForControlBlocks(SCL scl, CBCom cb
                             if (settings == null) {
                                 return newError(iedApLd, tControl, "Cannot configure communication for this ControlBlock because: No controlBlock communication settings found with these " + criteriaOrError.criteria);
                             }
-                            return configureControlBlockNetwork(scl.getCommunication(), settings, cbComSettings.appIdIterator, cbComSettings.macAddressIterator, tControl, iedApLd);
+                            return configureControlBlockNetwork(scl.getCommunication(), settings, appIdIterator, macIterator, tControl, iedApLd, appIdsAndMacsToReuse);
```
[the remark this answers is not in the document]
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java]: OK, well seen

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/util/Utils.java [hunk 5]
```diff
@@ -366,4 +327,14 @@ public static <T> T copySclElement(T object, Class<T> clazz) {
         }
     }
 
+    /**
+     * Extract content of P element of given type
+     *
+     * @param type type attribute of P element
+     * @param listOfP list of P elements
+     * @return content of P element if P element with given type exist, else empty Optional.
```
Comment (GuillaumeJAFFRE) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/util/Utils.java:335]: return VALUE of THE FIRST of TYPE element in listOfP if... ?

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java [hunk 6]
```diff
@@ -309,15 +312,44 @@ void configureNetworkForAllControlBlocks_should_create_SMV_elements() {
         MarshallerWrapper.assertValidateXmlSchema(scd);
     }
 
+    public static Stream<Arguments> ProvideCommunicationToReuse() {
+        TCommunication communicationWithDeletedCb = new TCommunication();
```
Comment (GuillaumeJAFFRE) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:316]: never used

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java [hunk 7]
```diff
@@ -309,15 +312,44 @@ void configureNetworkForAllControlBlocks_should_create_SMV_elements() {
...
+        tConnectedAP2.setApName("IED_NAME2");
+        tConnectedAP2.getGSE().add(newGse("LD_INST21", "CB_LD_INST21_GSI", "0000", "01-0C-CD-01-00-00"));
+        tConnectedAP2.getSMV().add(newSmv("LD_INST21", "CB_LD_INST21_SVI", "4000", "01-0C-CD-04-00-00"));
+        tSubNetwork2.getConnectedAP().add(tConnectedAP2);
+        communicationWithCommunicationToReuse.getSubNetwork().add(tSubNetwork2);
+
+        return Stream.of(
+                Arguments.of(named("no communication to reuse", null)),
+                Arguments.of(named("empty communication to reuse", new TCommunication())),
+                Arguments.of(named("communication to reuse with only deleted CB", communicationWithCommunicationToReuse)),
```
Comment (GuillaumeJAFFRE) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:339]: issue here ?

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java [hunk 8]
```diff
@@ -309,15 +312,44 @@ void configureNetworkForAllControlBlocks_should_create_SMV_elements() {
...
+                Arguments.of(named("communication to reuse with CB", null))
```
Comment (GuillaumeJAFFRE) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:340]: issue here ?

**Review by GuillaumeJAFFRE — COMMENTED** [review]
first read / not finished

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java [hunk 9]
```diff
@@ -388,14 +433,42 @@ public record Settings(Integer vlanId, Byte vlanPriority, TDurationInMilliSec mi
...
+     * @param mac   MAC-Address
+     */
+    record AppIdAndMac(long appId, long mac) {
+        static Optional<AppIdAndMac> from(TAddress address) {
+            if (address == null) {
+                return Optional.empty();
+            }
+            return Utils.extractFromP(APPID_P_TYPE, address.getP())
+                    .map(appId -> Integer.parseInt(appId, HEXADECIMAL_BASE))
+                    .flatMap(aapId -> Utils.extractFromP(MAC_ADDRESS_P_TYPE, address.getP())
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:469]:
```suggestion
                    .flatMap(appId -> Utils.extractFromP(MAC_ADDRESS_P_TYPE, address.getP())
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/api/ControlBlockEditor.java [hunk 10]
```diff
@@ -65,4 +66,22 @@ public interface ControlBlockEditor {
...
+     * else use range provided in cbCom.
+     * APPID and MAC-Address of communicationToReuse for ControlBlocks that exists in scd are excluded from the ranges given in cbCom.
+     *
+     * @param scd                  input SCD object. The object will be modified with the new GSE and SMV elements
+     * @param cbCom                communication settings to configure Control Block Communication
+     * @param communicationToReuse communication section to search for existing APPID and MAC-Address for ControlBlock
+     * @return list of encountered errors
+     * @see Utils#macAddressToLong(String) for the expected MAC address format
+     */
+    List<SclReportItem> configureNetworkForAllControlBlocks(SCL scd, CBCom cbCom, TCommunication communicationToReuse);
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/api/ControlBlockEditor.java:85]: Given that we simply require the Subnetwork list,why using  `TCommunication`  ? also `SCL` can be used as type

```suggestion
    List<SclReportItem> configureNetworkForAllControlBlocks(SCL scd, CBCom cbCom, SCL targetSCD);
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java [hunk 11]
```diff
@@ -89,18 +92,33 @@ public void removeAllControlBlocksAndDatasetsAndExtRefSrcBindings(final SCL scl)
...
+    }
+
+    private Stream<SclReportItem> configureNetworkForControlBlocks(SCL scl, Map<CbKey, AppIdAndMac> appIdsAndMacsToReuse, CBCom cbCom, TCBType tcbType) {
         CbComSettings cbComSettings;
         try {
             cbComSettings = parseCbCom(cbCom, tcbType);
         } catch (ScdException ex) {
             return Stream.of(SclReportItem.error("Control Block Communication setting files", ex.getMessage()));
         }
+        List<Long> reservedAppIds = appIdsAndMacsToReuse.values().stream().map(AppIdAndMac::appId).toList();
```
[the remark this answers is not in the document]
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:118]: I would prefer `appIdToReuse` and `macToReuse` eventually.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/util/Utils.java (hunk 5 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/util/Utils.java:335]: OK

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java (hunk 6 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:316]: It is used twice. Look at `@MethodSource` of `ParameterizedTest`.

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java (hunk 7 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:339]: yes. Fixed it

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java (hunk 8 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java:340]: yes. Fixed it.

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java (hunk 9 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java:469]: OK

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/api/ControlBlockEditor.java (hunk 10 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/api/ControlBlockEditor.java]: OK, I change to List of Subnetworks

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (GuillaumeJAFFRE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ControlBlockEditorService.java]: ok

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java (hunk 6 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/ControlBlockEditorServiceTest.java]: communicationWithDeletedCb has been removed.

**Pull request merged**